### PR TITLE
Honor `root_url` in core

### PR DESCRIPTION
**Pull request by danbernier**
@sigmavirus24 Here's a working solution for what I was proposing in #343.

To an end-user, the only effect is that the `:root_url` (and the `:protocol`) options now can be passed right to the PDFKit constructor. Configuring `root_url` still works, too.

This PR introduces `PDFKit::HtmlPreprocessor`, which could maybe also handle inserting the `<stylesheet>` tags, but I didn't want to change too much at once.

What do you think?

### CODE REVIEW

**Comment by danbernier** [comment]
Oh, I forgot to say that some of the commits are maybe excessively small. I committed a lot of the refactors individually, so you can see each of the moves in isolation. If you like the change, I can rebase & squash it all.

File: lib/pdfkit/html_preprocessor.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    def process(html, root_url, protocol)
+      html = translate_relative_paths(html, root_url) if root_url
+      html = translate_relative_protocols(html, protocol) if protocol
+      html
+    end
+
+    private
+
+    def translate_relative_paths(html, root_url)
+      # Try out this regexp using rubular http://rubular.com/r/hiAxBNX7KE
```
Comment (cdwort) [inline lib/pdfkit/html_preprocessor.rb]: Thanks for adding these!

Comment (cdwort) [inline lib/pdfkit/html_preprocessor.rb]: Errr, well, thanks for bringing them along with :P

**Comment by cdwort** [comment]
Hey @danbernier 

Thanks for the PR. I'll take a fuller look at it in the morning and @sigmavirus24 may have some thoughts before the.

I'm fine with small commits - we don't see too much activity in this codebase so the history doesn't need to be the most beautiful for debugging and public documentation. The Changelog generally covers each change anyways.

File: lib/pdfkit/html_preprocessor.rb (hunk 1 above)
Comment (danbernier) [inline lib/pdfkit/html_preprocessor.rb]: They're good comments, I'm all for keeping them :smile:

**Comment by danbernier** [comment]
@cdwort sounds good!

File: lib/pdfkit/html_preprocessor.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+class PDFKit
+  class HtmlPreprocessor
```
Comment (sigmavirus24) [inline lib/pdfkit/html_preprocessor.rb]: Why is this a class and not a module. There's nothing being stored here at all and there's no state.

File: lib/pdfkit/pdfkit.rb [hunk 3]
```diff
@@ -105,6 +108,13 @@ def style_tag_for(stylesheet)
     "<style>#{File.read(stylesheet)}</style>"
   end
 
+  def preprocess_html
+    if @source.html?
+      processed_html = PDFKit::HtmlPreprocessor.new.process(@source.to_s, @root_url, @protocol)
```
Comment (sigmavirus24) [inline lib/pdfkit/pdfkit.rb]: The `.new` here should be a hint that using a class is the wrong decision.

File: lib/pdfkit/html_preprocessor.rb (hunk 2 above)
Comment (sigmavirus24) [inline lib/pdfkit/html_preprocessor.rb]: Also the (unfortunate) existing style of this library is to use all caps in names, e.g., `PDFKit` so we should use `HTMLPreprocessor`.

**Comment by sigmavirus24** [comment]
I've restarted the two jobs that failed with errors. I wouldn't mind it if you squashed some of the history, but that's merely something that would be *nice to have* as opposed to a requirement for this to be merged.

File: lib/pdfkit/html_preprocessor.rb (hunk 2 above)
Comment (danbernier) [inline lib/pdfkit/html_preprocessor.rb]: Ok, I'll make it a module, & swap the name.

Comment (danbernier) [inline lib/pdfkit/html_preprocessor.rb]: If anyone ever decides to pull the stylesheet stuff into `HTMLPreprocessor`, maybe it'll have enough heft to be its own class. It'll be easy enough to change back at that point.

**Comment by danbernier** [comment]
@sigmavirus24 I made `HTMLProcessor` a module. If you're happy with this, I'm happy to squash it into 1 commit, before merging.

File: spec/html_preprocessor_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+require 'spec_helper'
+
+describe PDFKit::HTMLPreprocessor do
+  describe "#process" do
+    before do
+      @precprocessor = PDFKit::HTMLPreprocessor
```
Comment (cdwort) [inline spec/html_preprocessor_spec.rb]: Can we change `precprocessor` to `preprocessor`?

Comment (danbernier) [inline spec/html_preprocessor_spec.rb]: `Array.new(1000) { "D'oh!" }`

Fix coming.

File: spec/html_preprocessor_spec.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      expect(body).to eq("<html><head><link href='http://example.com/stylesheets/application.css' media='screen' rel='stylesheet' type='text/css' /></head><body><img alt='test' src=\"http://example.com/test.png\" /></body></html>")
+    end
+
+    it "correctly parses relative url with double quotes" do
+      @body = %{<link href="/stylesheets/application.css" media="screen" rel="stylesheet" type="text/css" />}
+      body = @precprocessor.process @body, @root_url, @protocol
+      expect(body).to eq("<link href=\"http://example.com/stylesheets/application.css\" media=\"screen\" rel=\"stylesheet\" type=\"text/css\" />")
+    end
+
+    it "correctly parses relative url with double quotes" do
```
Comment (cdwort) [inline spec/html_preprocessor_spec.rb]: Same test name as above. Were you going for special characters here?

Comment (danbernier) [inline spec/html_preprocessor_spec.rb]: Ahh, nice catch. I actually moved these from the Middleware spec, so I didn't focus on the descriptions so much. I'll clean them up.

Comment (danbernier) [inline spec/html_preprocessor_spec.rb]: I fixed up the descriptions, and that made me realize we were missing a test case - so I added that. :tada:

File: spec/html_preprocessor_spec.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
+require 'spec_helper'
+
+describe PDFKit::HTMLPreprocessor do
```
Comment (cdwort) [inline spec/html_preprocessor_spec.rb]: I would like to see another quick test for when root_url & protocol are nil. This would be the case for when PDFKit is not being used as middleware & the user doesn't explicitly set root_url & protocol. It's a simple test that prevents unintentional breakage in the future.

File: spec/html_preprocessor_spec.rb (hunk 5 above)
Comment (cdwort) [inline spec/html_preprocessor_spec.rb]: Hm, I see this was copied from the middleware tests. I'm guessing special characters?

**Comment by danbernier** [comment]
@cdwort I made the last 2 changes you suggested as rebase-fixup commits, so I'll definitely rebase this down before it's merged. Nice & tidy.

**Comment by cdwort** [comment]
Hey @danbernier Thanks for those. I appreciate you clearing up the test names even though that was what was there before. Obviously very helpful given that I was off-base on what the intent was to test there.

As I noted, I'd love to see one last addition to the tests for `HTMLPreprocessor`, which would be essentially the last test (or one with a URI), that sets `root_url` and `protocol` to nil. Unless I'm reading this wrong, that will be what typically happens when this library is used outside middleware.

Other than that, this is looking good! Thanks for the PR again.

**Comment by danbernier** [comment]
@cdwort No problem! I'm glad you caught those. I'm working on the tests for a nil `root_url` and/or `protocol` now (you're exactly right about when they'd be nil). Feels like the test suite is generally getting better :smile_cat:

**Comment by danbernier** [comment]
@cdwort I just pushed the specs for a nil `root_url` and `protocol`. Take a :eyes:

File: spec/html_preprocessor_spec.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        body = @preprocessor.process @body, @root_url, nil
+        expect(body).to eq(%{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='http://example.com/'><img src='http://example.com/logo.jpg'></a>})
+      end
+    end
+
+    context 'when root_url and protocol are both nil' do
+      it "returns the body safely, without interpolating" do
+        @body = %{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>}
+        body = @preprocessor.process @body, nil, nil
+        expect(body).to eq(%{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>})
```
Comment (cdwort) [inline spec/html_preprocessor_spec.rb]: There's really no need for `@body` to be a class var. Can we change it to `original_body`? (I know you just copied this from elsewhere that did that)

Changing that would allow us to make this test easier to grok, by making this:

```
expect(body).to eq original_body
```

**Comment by cdwort** [comment]
Yeah, agreed... it's always nice to see the test quite inch up in quality!

File: spec/html_preprocessor_spec.rb (hunk 7 above)
Comment (danbernier) [inline spec/html_preprocessor_spec.rb]: @cdwort I can do that. While I'm at it, what do you think about using `let` statements, instead of ivars, for the preprocessor, root_url, and protocol?

Comment (cdwort) [inline spec/html_preprocessor_spec.rb]: 100% :+1: if you don't mind doing it.

Comment (danbernier) [inline spec/html_preprocessor_spec.rb]: Nope, I don't mind.

It's done, and pushed.

File: spec/html_preprocessor_spec.rb [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        body = preprocessor.process original_body, root_url, nil
+        expect(body).to eq(%{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='http://example.com/'><img src='http://example.com/logo.jpg'></a>})
+      end
+    end
+
+    context 'when root_url and protocol are both nil' do
+      it "returns the body safely, without interpolating" do
+        original_body = %{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>}
+        body = preprocessor.process original_body, nil, nil
+        expect(body).to eq(%{<link href='//fonts.googleapis.com/css?family=Open+Sans:400,600' rel='stylesheet' type='text/css'><a href='/'><img src='/logo.jpg'></a>})
```
Comment (cdwort) [inline spec/html_preprocessor_spec.rb]: I'd still personally prefer to see the expectation written 

```
expect(body).to eq original_body
```

because then I don't have to stare at both lines to make sure they're the same.

Other than that, this looks good!

Comment (danbernier) [inline spec/html_preprocessor_spec.rb]: Ahhhh, I totally spaced out there. You're right. Hang on...

**Comment by cdwort** [comment]
This is looking good. I'll rerun the tests until they pass.

I'll wait until this evening to give anyone from @pdfkit/rb a chance to take a look, but I don't see any problems merging this.

Before this goes out, README.md documentation will need to be updated: https://github.com/pdfkit/pdfkit/blob/master/README.md#middleware

@danbernier If you'd like to take a quick stab at that, then I'm guessing we can cut a release soon. Otherwise I'll try to get to it soon.

File: spec/html_preprocessor_spec.rb (hunk 8 above)
Comment (danbernier) [inline spec/html_preprocessor_spec.rb]: done.

**Comment by danbernier** [comment]
@cdwort I'll draft a README update, and see how it goes. Since it seems the code changes are done, I'll squash the code changes down, too.

**Comment by danbernier** [comment]
@cdwort I updated the README, though it could benefit from another set of eyes.

File: spec/html_preprocessor_spec.rb (hunk 6 above)
Comment (sigmavirus24) [inline spec/html_preprocessor_spec.rb]: :+1: on the test.

Comment (cdwort) [inline spec/html_preprocessor_spec.rb]: Dan added that test and a few other test cases. I consider this to be addressed.

Comment (sigmavirus24) [inline spec/html_preprocessor_spec.rb]: Ah, I hadn't looked too closely. I was relying on GitHub to hide this when it was completed. I'm +1 on this PR at this point then. Feel free to merge @cdwort

**Comment by cdwort** [comment]
I can't get a clean run of this on Rubies 1.9.3 and 2.1, but it's failing because it can't install PDFKit.

**Comment by danbernier** [comment]
Yeah, I saw that in the builds, and wasn't sure what to make of it. Is it
something unique to this branch, or have you seen it before?

~
This message was made with 100% thumbs.

**Comment by cdwort** [comment]
No, @danbernier, I don't think it's this branch. I think it's a bundler bug here: https://github.com/bundler/bundler/issues/3558

**Comment by danbernier** [comment]
Any updates on the bundler bug, or this PR?

**Comment by cdwort** [comment]
The 1.9.3 run uses Bundles 1.11.1, and all the other builds use Bundler 1.7.6, so something strange is going on here.

**Comment by danbernier** [comment]
Is there anything I can do to help?

**Comment by cdwort** [comment]
Hey @danbernier Just a quick status update. I'm struggling to get ruby 1.9.3 working on my personal computer so I can run tests. I'll keep letting you know where we are.

**Comment by danbernier** [comment]
@cdwort ok. FWIW, when I run the specs under both 1.9.3 and 2.1.0, they pass for me. I know that's not much help =/

**Comment by cdwort** [comment]
Yeah, I'm going to try again tonight... now that I can't use that machine for my job I've forgotten the quirks. It's a fine laptop, just hasn't been used heavily in a bit.

If I can't get it working, I'll trust you on this. It's the 1.9.3 that worries me a tiny bit.. the differences between the 2.0 and 2.1 are minimal.

**Comment by danbernier** [comment]
@cdwort any luck?

**Comment by cdwort** [comment]
Hey @danbernier Can you pull in master for the travis fix?

**Comment by danbernier** [comment]
@cdwort done!

**Comment by danbernier** [comment]
@cdwort :tada: Travis says it's all green

**Comment by danbernier** [comment]
@cdwort @sigmavirus24 thanks for all your help!
